File: Python/sources/Functional programming/mapping.py
from numpy import array, matmul


def f(x): 
    return x**2 

def g(x,n): 
    return x**n 

class tensor(): 
    def __init__(self, L):
        self.x = L 

def square(A):
    B = tensor(A)
    B.x =  matmul( A.x, A.x ) 
    return B 

def mapping_example(): 

# mapping with lists 
  x = [ 1, 2, 3, 4 ]

  y1 = list( map(f, x) ) 
  

  y2 = list( map( lambda x : g(x, 2), x ) ) 

  print("y1=", y1) 
  print("y2=", y2) 

# mapping with arrays 
  x = array( [ 1, 2, 3, 4] )
  y = f ( x )
  print("y=", y) 

# mappings with tensors 
  A  = tensor( array([ [ 1, 2], [3,4] ]) )
  print("A=",A.x) 

  B = square(A)
  print(" A*A =", B.x[1,:] )  
  
# mapping of tensors elementwise operation  
  C = array( [ A, A, A] )
  print(" C =", C[2].x[1,:] ) 
  #D = square(C) ! it does not work 
  D = array( list( map(square, C)) ) 
  print( "A*A=", D[2].x[1, :] ) 

  A  = array([ [ 1, 2], [3,4] ])
  C = array( [A,A,A] )
  D = matmul( C, C ) 
  print( "C*C=", D[2, 1, :] ) 

File: Fortran/sources/Advanced_programming/functional programming/mapping.f90
module mapping 
    
    implicit none 
    
    type tensor
        real, allocatable :: x(:,:) 
    end type 
    
    
contains     
    
elemental real function f(x)
    real, intent(in) :: x 
    f = x**2 
end function 



elemental function square(A) result(F) 
    type(tensor), intent(in) :: A
    type (tensor) :: F
    
    F % x = matmul( A % x, A % x ) 
    
end function 

elemental real function g(x, n)
    real, intent(in) :: x 
    integer, intent(in) :: n 
    g = x**n 
end function 


subroutine mapping_example 


real :: x(4) = [ 1, 2, 3, 4 ]
real :: y1(4), y2(4) 
type (tensor) :: A, B, C(3), D(3) 
integer :: i 


! Python mapping: y1 = list( map(f, x) ) 
! Fortran: 
  y1 = f(x) 

! Python mapping: y2 = list( map( lambda x : g(x, 2), x ) ) 
! Fortran:   
  y2 = g(x, 2) 

  write(*, '(A3, 10es15.5)') "y1=", y1
  write(*, '(A3, 10es15.5)') "y2=", y2

! Tensor B = A * A   
  allocate ( A % x(2,2), B % x(2,2) ) 
  do i=1, 3 
     allocate( C(i) % x(2,2) ) 
  end do 
 
 A % x(1,:) = [ 1, 2 ]
 A % x(2,:) = [ 3, 4 ] 
 C = A 
 B = square(A)
 write(*, '(A3, 10es15.5)') "A*A=", B % x(1, :) 
  
! mapping of tensors elementwise operation  
 D = square(C)
 write(*, '(A3, 10es15.5)') "A*A=", D(3) % x(1, :) 
 
  
end subroutine 

end module 
